### DISCUSSION TITLE
Integrate riscv-none-elf-gcc into Arduino IDE?

### DISCUSSION THREAD

**Discussion by ilg-ul** [xpack-dev-tools/riscv-none-elf-gcc-xpack#3] [General]
Frederic @fpistm, do you know how difficult would be to integrate riscv-none-elf-gcc into Arduino IDE for use with RISC-V projects?

I know you did a similar thing with arm-none-eabi-gcc, and perhaps we can do the same for the RISC-V community?

**Comment by ilg-ul** [comment]
I mentioned this on the RISC-V SW forum:

- <https://groups.google.com/a/groups.riscv.org/g/sw-dev/c/_AtIOOMX1Os>

**Reply by ilg-ul** [reply]
As mentioned in the RISC-V forum, there is package for the SiFive boards:

- <https://github.com/sifive/cinco>

**Reply by fpistm** [reply]
About cinco repo, it provide json file, so in that case the [JSON ](http://static.dev.sifive.com/bsp/arduino/package_sifive_index.json) could reference the toolchain and the core have to simply use it instead of the one it provides: https://github.com/sifive/cinco/blob/master/hardware/freedom_e/platform.txt#L7

**Reply by ilg-ul** [reply]
I would like to not interfere with their packages, but offer an alternate solutions, for those who prefer a newer toolchain.

In other words, to provide a package to be installed along the cinco one, and the users have the freedom to choose between multiple toolchains.

**Reply by fpistm** [reply]
AFAIK it is not possible to provide separate toolchain package and choose which one to use. 
As stated above only way I saw is to provide an another json with custom packages: 
https://github.com/xpack-dev-tools/riscv-none-elf-gcc-xpack/discussions/3#discussioncomment-2774580

But I'm  not familiar with this. I would recommend to contact cinco maintainer's to find the best way if they are interested.

**Comment by fpistm** [comment]
Hi @ilg-ul 
The way to specify Arduino which tools need to be installed are detailed here:
https://arduino.github.io/arduino-cli/0.22/package_index_json-specification/


For STM32duino, we define which tools version we need:
https://github.com/stm32duino/BoardManagerFiles/blob/6e08c29204043dadb17c36ca32ae5680f178f6af/package_stmicroelectronics_index.json#L167-L172

And it is described later here:
https://github.com/stm32duino/BoardManagerFiles/blob/main/package_stmicroelectronics_index.json#L311-L359

So to resume an Arduino core designed for riscv arch have to provide its own package_index.json with the correct information.

**Reply by ilg-ul** [reply]
Thank you Frederic, 

That's a bit too much info for me to grasp at once, but it is a good reference.

> an Arduino core designed for riscv arch have to provide its own package_index.json with the correct information.

I don't know if there is an _Arduino core designed for RISC-V_, since I don't know how to search for repositories of Arduino packages.

**Comment by ilg-ul** [comment]
Is it possible to define an Arduino package with only the toolchain? 

If so, how to add a link to it from some public Arduino site, so that people will know about it?

**Comment by fpistm** [comment]
Toolchain is linked to a core architecture, so if there is no core for riscv no need to have toolchain.

Here a list of unofficial 3rd party cores:
https://github.com/arduino/Arduino/wiki/Unofficial-list-of-3rd-party-boards-support-urls

But it is not exhaustive.

**Reply by ilg-ul** [reply]
> Toolchain is linked to a core architecture, 

ok

> ... so if there is no core for riscv no need to have toolchain.

I'm not sure exactly what a _core_ is, but there is at least one RISC-V package, mentioned at <https://github.com/sifive/cinco>.

To rephrase the question, if there are third-party RISC-V package, with their own toolchains, is it possible to define a separate package with only the xPack riscv-none-elf-gcc toolchain, and people to install it along with their other RISC-V packages and be able to switch to this new toolchain?

**Reply by fpistm** [reply]
> I'm not sure exactly what a _core_ is,

https://docs.arduino.cc/learn/starting-guide/cores

Extract:

> Cores are necessary to make new microcontrollers compatible with your Arduino Software (IDE) and, possibly, the existing sketches and libraries. We develop the cores for the new microcontrollers used in our new generation boards, but anyone may develop a core for their own boards following the rules and requirements we have issued.




> To rephrase the question, if there are third-party RISC-V package, with their own toolchains, is it possible to define a separate package with only the xPack riscv-none-elf-gcc toolchain, and people to install it along with their other RISC-V packages and be able to switch to this new toolchain?

I saw, It is possible to use another core to create new one with some custom but I never tested and don't knwo more:
https://arduino.github.io/arduino-cli/0.19/platform-specification/#referencing-another-core-variant-or-tool

**Reply by fpistm** [reply]
Your best bet is to open an issue to cinco repo to ask tem if they want to integrate this toolchain.

**Reply by ilg-ul** [reply]
> open an issue to cinco repo

The last commit there is from Sep 2019. They discontinued support for Arduino.

**Comment by ilg-ul** [comment]
So, to conclude, the Arduino design does not allow to define separate toolchain packages, they need to be directly defined in core packages.

In other words, in order to use a different toolchain, core packages maintainers should update their definitions.

If anyone will be interested in this, they are welcomed to use the toolchain, but I personally don't use Arduino and I'm not familiar with the details.
